src/Microsoft.Azure.WebJobs.Extensions.Kafka/Listeners/KafkaTopicScaler.cs

```diff
@@ -0,0 +1,255 @@
+            }
+
+            return new List<TopicPartition>();
+        }
+
+        async Task<ScaleMetrics> IScaleMonitor.GetMetricsAsync()
+        {
+            return await GetMetricsAsync();
+        }
+
+        public Task<KafkaTriggerMetrics> GetMetricsAsync()
+        {
+            var operationTimeout = TimeSpan.FromSeconds(5);
+            var allPartitions = topicPartitions.Value;
+            if (allPartitions == null)
+            {
+                return Task.FromResult(new KafkaTriggerMetrics(0L, 0));
+            }
+
+            var ownedCommittedOffset = consumer.Committed(allPartitions, operationTimeout);
+            var partitionWithHighestLag = Partition.Any;
+            long highestPartitionLag = 0L;
+            long totalLag = 0L;
+            foreach (var topicPartition in allPartitions)
+            {
+                // This call goes to the server always which probably yields the most accurate results. It blocks.
+                // Alternatively we could use consumer.GetWatermarkOffsets() that returns cached values, without blocking.
+                var watermark = consumer.QueryWatermarkOffsets(topicPartition, operationTimeout);
+
+                var commited = ownedCommittedOffset.FirstOrDefault(x => x.Partition == topicPartition.Partition);
+                if (commited != null)
+                {
+                    long diff;
+                    if (commited.Offset == Offset.Unset)
+                    {
+                        diff = watermark.High.Value;
+                    }
+                    else
+                    {
+                        diff = watermark.High.Value - commited.Offset.Value;
+                    }
+
+                    totalLag += diff;
+
+                    if (diff > highestPartitionLag)
+                    {
+                        highestPartitionLag = diff;
+                        partitionWithHighestLag = topicPartition.Partition;
+                    }
+                }
+            }
+
+            if (partitionWithHighestLag != Partition.Any)
+            {
+                logger.LogInformation("Total lag in '{topic}' is {totalLag}, highest partition lag found in {partition} with value of {offsetDifference}", this.topicName, totalLag, partitionWithHighestLag.Value, highestPartitionLag);
+            }
+
+            return Task.FromResult(new KafkaTriggerMetrics(totalLag, allPartitions.Count));
+        }
+
+        public ScaleStatus GetScaleStatus(ScaleStatusContext context)
+        {
+            return GetScaleStatusCore(context.WorkerCount, context.Metrics?.OfType<KafkaTriggerMetrics>().ToArray());
+        }
+
+        public ScaleStatus GetScaleStatus(ScaleStatusContext<KafkaTriggerMetrics> context)
+        {
+            return GetScaleStatusCore(context.WorkerCount, context.Metrics?.ToArray());
+        }
+
+        private ScaleStatus GetScaleStatusCore(int workerCount, KafkaTriggerMetrics[] metrics)
+        {
```

Comment:
You need a null/empty check here on metrics, since your calls above can evaluate to null